### DISCUSSION TITLE
WinUI3 / AppSDK - PenButtonListener

### DISCUSSION THREAD

**Discussion by onslauth** [microsoft/WindowsAppSDK#4022] [Q&A] [answered]
How do you get an instance of PenButtonListener in AppSDK? I am trying to listen for the TailButtonClicked event in a global context but cannot find a way to receive the event.

**Comment by DarranRowe** [comment]
There doesn't seem to be a Windows App SDK version of this class.
This shouldn't stop you from using the UWP version though. From what I can see in the documentation.
1) Be sure that you are running Windows 10 2004/20H1 or newer. If your application is packaged and you have set the minimum version to Windows 10 2004/20H1 (19041) then you can just assume that this is true. If the minimum version is lower than this, or your application is unpackaged then you should use [Windows.Foundation.Metadata.ApiInformation](https://learn.microsoft.com/en-us/uwp/api/windows.foundation.metadata.apiinformation?view=winrt-22621) and check to see if the Windows.Foundation.UniversalApiContract contract v10 is available. This contract and version is documented in the [PenButtonListener](https://learn.microsoft.com/en-us/uwp/api/windows.devices.input.penbuttonlistener?view=winrt-22621#windows-requirements) documentation.
2) Call [PenButtonListener.GetDefault](https://learn.microsoft.com/en-us/uwp/api/windows.devices.input.penbuttonlistener.getdefault?view=winrt-22621#windows-devices-input-penbuttonlistener-getdefault) to get the instance of PenButtonListener.
3) Hook up the events using the instance that was returned. Also, remember to keep the instance of this class alive.

**Reply by onslauth** [reply]
Thank you for the detailed reply. I have already tried the `PenButtonListener.GetDefault` method, and after assigning a listener to the `TailButtonClicked`, the event it is never fired.

Here is the example code:
```
        public TestPage( )
        {
            var val = ApiInformation.IsApiContractPresent("Windows.Foundation.UniversalApiContract", 10);
            if (val == true)
            {
                var pl = PenButtonListener.GetDefault();
                pl.TailButtonClicked += Pl_TailButtonClicked;
            }
        }
    
        // This function is never reached.
        private void Pl_TailButtonClicked(PenButtonListener sender, PenTailButtonClickedEventArgs args)
        {
            throw new NotImplementedException();
        }
```

I've added breakpoints and stepped through the code to verify that it is working and it enters the if statement. However the `Pl_TailButtonClicked` function is never executed.

I have also verified that the Pen device is paired and inside `Settings > Bluetooth & devices > Pen & Windows Ink` settings page have changed the shortcut button to various values and confirmed they work correctly, and `Allow apps to override the shortcut button behaviour` is enabled.

Is there perhaps anything else you can think of to try?

**Reply by DarranRowe** [reply, marked as answer]
Does IsSupported return true or false? If it is unsupported, then I would guess that it is just not supported in desktop applications, especially if it works in a UWP application.
Annoyingly, the Windows Runtime documentation can be like this. There is no indicator as to whether it is supposed to be supported.

**Reply by onslauth** [reply]
Thank you very much for all the help.

`IsSupported` returns false.
